tools/hsettp/kvs-cn-tree-get.c

```diff
@@ -327,42 +324,44 @@ kvs_cn_tree_get_parse_values(cJSON *const body, int *const len, char ***const va
-                    const char *job_rule = "-";
-                    const char *job_wmesg = "-";
+                entry = cJSON_GetObjectItemCaseSensitive(k, "job");
+                assert(cJSON_IsNull(entry) || cJSON_IsObject(entry));
+                if (cJSON_IsObject(entry)) {
+                    uint id;
+                    ulong time;
+                    const char *rule;
+                    const char *wmesg;
+                    const char *action;
 
-                    job_id = cJSON_GetNumberValue(entry);
+                    entry = cJSON_GetObjectItemCaseSensitive(k, "id");
+                    assert(cJSON_IsNumber(entry));
+                    id = cJSON_GetNumberValue(entry);
 
-                    entry = cJSON_GetObjectItemCaseSensitive(k, "job_action");
-                    if (cJSON_IsString(entry))
-                        job_action = entry->valuestring;
+                    entry = cJSON_GetObjectItemCaseSensitive(k, "action");
+                    assert(cJSON_IsString(entry));
+                    action = cJSON_GetStringValue(entry);
 
-                    entry = cJSON_GetObjectItemCaseSensitive(k, "job_rule");
-                    if (cJSON_IsString(entry))
-                        job_rule = entry->valuestring;
+                    entry = cJSON_GetObjectItemCaseSensitive(k, "rule");
+                    assert(cJSON_IsString(entry));
+                    rule = cJSON_GetStringValue(entry);
 
-                    entry = cJSON_GetObjectItemCaseSensitive(k, "job_wmesg");
-                    if (cJSON_IsString(entry))
-                        job_wmesg = entry->valuestring;
+                    entry = cJSON_GetObjectItemCaseSensitive(k, "wmesg");
+                    assert(cJSON_IsString(entry));
+                    wmesg = cJSON_GetStringValue(entry);
 
-                    entry = cJSON_GetObjectItemCaseSensitive(k, "job_time");
-                    if (cJSON_IsNumber(id))
-                        job_time = cJSON_GetNumberValue(entry);
+                    entry = cJSON_GetObjectItemCaseSensitive(k, "time");
+                    assert(cJSON_IsNumber(entry));
+                    time = cJSON_GetNumberValue(entry);
 
                     rc = snprintf(buf, sizeof(buf), "- %u %s/%s %lu:%02lu %s",
-                                  job_id,
-                                  job_action,
-                                  job_rule,
-                                  (job_time / 60) % 60, job_time % 60,
-                                  job_wmesg);
-                    if (rc > 0)
-                        (*values)[offset] = strdup(buf);
-                }
-
-                if (!(*values)[offset])
+                        id, action, rule, (time / 60) % 60, time % 60, wmesg);
+                    assert(rc > 0);
+                    (*values)[offset] = strdup(buf);
+                    if (!(*values)[offset])
+                        return merr(ENOMEM);
+                } else {
```

Comment:
Much less resilient now...

Reply:
What do you mean?

Reply:
All these asserts are really starting to bug me.  I don't see any reason this parser cannot gracefully handle each case, the output isn't so critical that we need to crash the process should we receive unexpected input.  Best case in a release build is that we segfault, worst case we end up corrupting the user's data.

IMO, I would much rather a user come to us and say, "I ran this command and got some results that don't look right", vs "I ran this command and it crashed my app".

https://en.wikipedia.org/wiki/Robustness_principle

Reply:
This is in the `hsetpp` tool, not `libhse`.  So we won't be corrupting KVDB.

Reply:
Pretty much all the asserts in this file are to cover developer error like the JSON schema of a response. Not sure what you mean. If these asserts aren't here and somethings gets messed up, you will most definitely be entering undefined behavior.

Reply:
In the case you are out of memory, you won't be seeing any data either because the application will exit with an ENOMEM and will not recover.

Reply:
This whole tool is kind of brittle in that it requires the developer to keep **it** and the docs in lock step, so the asserts exist to tell you what you have done wrong. And during the course of development of this tool, I found the asserts extremely useful.

Reply:
Sorry, was thinking this was the rest hook.  Anyway, you can leave the assert and handle it so that the tool in a release build doesn't become useless because of a bug you didn't anticipate.

Reply:
Also adding proper tests for this tool would go a long way in catching these issues before someone ran into issues in release builds.